src/bloom.cpp

```diff
@@ -20,33 +20,61 @@
 
 #define LN2SQUARED 0.4804530139182014246671025263266649717305529515945455
 #define LN2 0.6931471805599453094172321214581765680755001343602552
+#define MIN_N_HASH_FUNC 1
 
 using namespace std;
 
 void CBloomFilter::setup(unsigned int nElements,
     double nFPRate,
     unsigned int nTweakIn,
     unsigned char nFlagsIn,
-    bool size_constrained,
+    bool fSizeConstrained,
     uint32_t nMaxFilterSize = SMALLEST_MAX_BLOOM_FILTER_SIZE)
 {
     if (nElements == 0)
     {
         LOGA("Construction of empty CBloomFilter attempted.\n");
         nElements = 1;
     }
-    unsigned desired_vdata_size = (unsigned int)(-1 / LN2SQUARED * nElements * log(nFPRate) / 8);
+    unsigned int nDesiredSize = (unsigned int)(-1 / LN2SQUARED * nElements * log(nFPRate) / 8);
 
-    if (size_constrained)
-        desired_vdata_size = min(desired_vdata_size, nMaxFilterSize);
+    if (fSizeConstrained)
+        nDesiredSize = min(nDesiredSize, nMaxFilterSize);
 
-    vData.resize(desired_vdata_size, 0);
+    vData.resize(nDesiredSize, 0);
     isFull = vData.size() == 0;
     isEmpty = true;
 
     nHashFuncs = (unsigned int)(vData.size() * 8 / nElements * LN2);
 
-    if (size_constrained)
+    if (fSizeConstrained)
+        nHashFuncs = min(nHashFuncs, MAX_HASH_FUNCS);
+
+    nTweak = nTweakIn;
+    nFlags = nFlagsIn;
+}
+
+void CBloomFilter::setupGuaranteeFPR(unsigned int nElements,
```

Comment:
Actually, I intentionally duplicated the setup method (modulo edits) in order to avoid having to make subtle changes to the existing `CBloomFilter` implementation. Originally @ptschip suggested subclassing `CBloomFilter`, but I needed access to private members, so we settled on this approach.

The fact that `nDesiredSize` can be arbitrarily large is not an accident. I need the Bloom filter to honor the requested false positive rate. I don't think that there is potential for data-blowup here because the entire `vData` vector is deserialized, so any filter that would be huge in memory would already be huge on the wire. With that being said, do we already avoid deserializing really huge messages into memory?

Reply:
Except for the almost duplication, the issue I have regarding `fSizeConstrained` is that you take the flag in the constructor, but only seem to honor it partially. `nDesiredSize` is unlimited, but `nHashFuncs` is still being limited. Is that really intentional? Assuming you want an exact FPR, isn't the whole notion of having `fSizeConstrained` kind of counterproductive?